hedera-node/hedera-app/src/main/java/com/hedera/node/app/signature/AppSignatureVerifier.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.signature;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.Key;
+import com.hedera.hapi.node.base.SignatureMap;
+import com.hedera.node.app.spi.signatures.SignatureVerifier;
+import com.hedera.node.app.spi.signatures.SimpleKeyCount;
+import com.hedera.node.app.spi.signatures.SimpleKeyVerification;
+import com.hedera.node.config.data.HederaConfig;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.function.Function;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Implements the {@link SignatureVerifier} interface for verifying signatures using the same components
+ * as used by the app workflows.
+ */
+@Singleton
+public class AppSignatureVerifier implements SignatureVerifier {
+    private static final int EDDSA_COUNT_INDEX = 0;
+    private static final int ECDSA_COUNT_INDEX = 1;
+
+    private final HederaConfig hederaConfig;
+    private final SignatureExpander signatureExpander;
+    private final com.hedera.node.app.signature.SignatureVerifier signatureVerifier;
+
+    @Inject
+    public AppSignatureVerifier(
+            @NonNull final HederaConfig hederaConfig,
+            @NonNull final SignatureExpander signatureExpander,
+            @NonNull final com.hedera.node.app.signature.SignatureVerifier signatureVerifier) {
+        this.hederaConfig = requireNonNull(hederaConfig);
+        this.signatureExpander = requireNonNull(signatureExpander);
+        this.signatureVerifier = requireNonNull(signatureVerifier);
+    }
+
+    @Override
+    public boolean verifySignature(
+            @NonNull final Key key,
+            @NonNull final Bytes bytes,
+            @NonNull final SignatureMap signatureMap,
+            @Nullable final Function<Key, SimpleKeyVerification> simpleKeyVerifier) {
+        final Set<ExpandedSignaturePair> sigPairs = new HashSet<>();
+        signatureExpander.expand(key, signatureMap.sigPair(), sigPairs);
+        final var results = signatureVerifier.verify(bytes, sigPairs);
```

Comment:
If I start here at the call to `verify` it takes me to the implementation of `super`'s `verify` which is `SignatureVerifierImpl.verify` and there at L73 for an EC signature it calls `createPreparerForEC` which at L110 does a `keccak256DigestOf` the incoming bytes.  

But up above at `SignatureVerifier` the class comment says that for EC the incoming bytes is already the keccak hash of the payload.  (And that's the way the HAS system contract methods are defined, too: `isAuthorizedRaw` and `isAuthorized` are defined in HIP-632 to take a `messageHash`.

So is that 2nd keccak hash one too many?

Reply:
Excellent question...we have to sort out what bytes people will be passing to this system contract...sounds like based on `isAuthorizedRaw` we want the bytes here to be a 32-byte hash only?

Reply:
Yes I think that's ok.  Well, for EC.  I think for ED, based on what I found on the intertubes, signatures are most often done with either SHA-384 or SHA-512 which of course have different lengths.